webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/domain/reports/StatusReports.scala

```diff
@@ -629,3 +627,331 @@ object NodeStatusReportSerialization {
+      @jsonField("nid")
+      nodeId:         NodeId,
+      @jsonField("rid")
+      ruleId:         RuleId,
+      @jsonField("ct")
+      complianceTag:  PolicyTypeName,
+      @jsonField("art")
+      agentRunTime:   Option[DateTime],
+      @jsonField("cid")
+      configId:       Option[NodeConfigId],
+      @jsonField("exp")
+      expirationDate: DateTime,
+      @jsonField("dsrs")
+      directives:     Seq[JDirectiveStatusReport]
+  ) {
+    def to: RuleNodeStatusReport = this.transformInto[RuleNodeStatusReport]
+  }
+
+  object JRuleNodeStatusReport {
+    implicit val a: Transformer[JRuleNodeStatusReport, RuleNodeStatusReport] = {
+      Transformer
+        .define[JRuleNodeStatusReport, RuleNodeStatusReport]
+        .withFieldComputed(_.directives, _.directives.map(d => (d.directiveId, d.transformInto[DirectiveStatusReport])).toMap)
+        .buildTransformer
+    }
+    implicit val b: Transformer[RuleNodeStatusReport, JRuleNodeStatusReport] = {
+      Transformer
+        .define[RuleNodeStatusReport, JRuleNodeStatusReport]
+        .withFieldComputed(
+          _.directives,
+          _.directives.map { case (_, d) => d.transformInto[JDirectiveStatusReport] }.toSeq.sortBy(_.directiveId.uid.value)
+        )
+        .buildTransformer
+    }
+
+    implicit val codecJRuleNodeStatusReport: JsonCodec[JRuleNodeStatusReport] = DeriveJsonCodec.gen
+
+    def from(r: RuleNodeStatusReport): JRuleNodeStatusReport = {
+      r.transformInto[JRuleNodeStatusReport]
+    }
+  }
+
+  @jsonHint("dsr")
+  final case class JDirectiveStatusReport(
+      @jsonField("did")
+      directiveId: DirectiveId,
+      @jsonField("pts")
+      policyTypes: PolicyTypes,
+      @jsonField("csrs")
+      components:  List[JComponentStatusReport]
+  ) {
+    def to: DirectiveStatusReport = this.transformInto[DirectiveStatusReport]
+  }
+
+  object JDirectiveStatusReport {
+    implicit lazy val i: Iso[JDirectiveStatusReport, DirectiveStatusReport] = Iso.derive
+
+    implicit val codecJDirectiveStatusReport: JsonCodec[JDirectiveStatusReport] = DeriveJsonCodec.gen
+
+    def from(r: DirectiveStatusReport): JDirectiveStatusReport = {
+      r.transformInto[JDirectiveStatusReport]
+    }
+  }
+
+  sealed trait JComponentStatusReport {
+    def componentName: String
+    def to:            ComponentStatusReport = this.transformInto[ComponentStatusReport]
+  }
+
+  object JComponentStatusReport {
+    // can't use Iso and need to teach chimey about recursive sealed coproduct
+    implicit lazy val a1: Transformer[JBlockStatusReport, BlockStatusReport]         = Transformer.derive
+    implicit lazy val b1: Transformer[JValueStatusReport, ValueStatusReport]         = Transformer.derive
+    implicit lazy val c1: Transformer[JComponentStatusReport, ComponentStatusReport] = {
+      case x: JBlockStatusReport => x.transformInto[BlockStatusReport]
+      case x: JValueStatusReport => x.transformInto[ValueStatusReport]
+    }
```

Comment:
Ah, I saw the coproduct thing but I didn't find how to make it compile. 
I prefer to keep the name disctincts, because it's already hard enought to see what goes where, with in addition the JsonCodec one, the imports with risk of silent overriding, etc.